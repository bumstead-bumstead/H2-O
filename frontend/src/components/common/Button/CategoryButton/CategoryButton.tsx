import type { ButtonHTMLAttributes } from 'react';
import styled from '@emotion/styled';

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  isClicked?: boolean;
}

function CategoryButton({ children, isClicked = false, ...restProps }: Props) {
  return (
    <StyledButton isClicked={isClicked} {...restProps}>
      {children}
    </StyledButton>
  );
}

export default CategoryButton;

const StyledButton = styled.button<Props>`
  ${({ theme }) => theme.typography.TextKRMedium14}
  background-color: ${({ theme, isClicked }) => (isClicked ? theme.colors.skyBlueCardBg : theme.colors.white)};
  color: ${({ theme, isClicked }) => (isClicked ? theme.colors.primary500 : theme.colors.gray600)};
  padding: 6px 20px;
  border: 1px solid ${({ theme, isClicked }) => (isClicked ? theme.colors.skyBlue : theme.colors.gray200)};
  border-radius: 20px;

  &:hover {
    color: ${({ theme }) => theme.colors.primary500};
    border: 1px solid ${({ theme }) => theme.colors.skyBlue};
    background-color: ${({ theme, isClicked }) => !isClicked && theme.colors.white};
  }

  &:active {
    background-color: ${({ theme }) => theme.colors.primary100};
  }
`;
