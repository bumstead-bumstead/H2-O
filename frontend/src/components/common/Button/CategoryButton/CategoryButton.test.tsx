import CategoryButton from './CategoryButton';
import { render, screen, userEvent, waitFor } from '@/tests/test-util';

describe('CTAButton 컴포넌트 테스트', () => {
  it('버튼이 정상적으로 렌더링된다.', () => {
    render(<CategoryButton>버튼</CategoryButton>);

    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('버튼의 onClick 핸들러가 클릭시 정상적으로 호출된다.', async () => {
    const mockOnClick = vi.fn();
    render(<CategoryButton onClick={mockOnClick}>버튼</CategoryButton>);

    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(mockOnClick).toBeCalledTimes(1));
  });

  it('버튼에 disabled 속성을 전달하면 disabled 된다.', async () => {
    const mockOnClick = vi.fn();
    render(
      <CategoryButton disabled onClick={mockOnClick}>
        버튼
      </CategoryButton>,
    );

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  describe('CategoryButton 스냅샷 테스트', () => {
    it('isClicked속성이 false일 때 스냅샷이 일치한다.', () => {
      render(<CategoryButton isClicked={false}>버튼</CategoryButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('isClicked속성이 true일 때 스냅샷이 일치한다.', () => {
      render(<CategoryButton isClicked>버튼</CategoryButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });
  });
});
