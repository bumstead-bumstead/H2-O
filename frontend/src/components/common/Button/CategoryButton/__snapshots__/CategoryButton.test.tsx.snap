// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CTAButton 컴포넌트 테스트 > CategoryButton 스냅샷 테스트 > isClicked속성이 false일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  background-color: #FFFFFF;
  color: #838588;
  padding: 6px 20px;
  border: 1px solid #DADCDD;
  border-radius: 20px;
}

.emotion-0:hover {
  color: #00397B;
  border: 1px solid #A2C7E7;
  background-color: #FFFFFF;
}

.emotion-0:active {
  background-color: #CCD7E5;
}

<button
  class="emotion-0"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CategoryButton 스냅샷 테스트 > isClicked속성이 true일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  background-color: #A2C7E74D;
  color: #00397B;
  padding: 6px 20px;
  border: 1px solid #A2C7E7;
  border-radius: 20px;
}

.emotion-0:hover {
  color: #00397B;
  border: 1px solid #A2C7E7;
}

.emotion-0:active {
  background-color: #CCD7E5;
}

<button
  class="emotion-0"
>
  버튼
</button>
`;
